client/OPUNetGameSelectWnd.cpp

```diff
@@ -469,171 +468,184 @@ void OPUNetGameSelectWnd::OnReceive(Packet &packet)
+{
+	if (!ReceiveJoin(packet)) {
+		return; // Discard packet
 	}
-	case tlcJoinGranted:		// [Fall through]
-	case tlcJoinRefused:
-		// Verify packet size
-		if (packet.header.sizeOfPayload != sizeof(JoinReply)) {
-			return;						// Discard packet
-		}
-		// Make sure we've requested to join a game
-		if (joiningGame == nullptr)
-		{
-			// **DEBUG**
-			SetStatusText("Unexpected Join reply received");
-			return;						// Discard packet
-		}
-		// Check the session identifier
-		if (packet.tlMessage.joinReply.sessionIdentifier != joiningGame->sessionIdentifier)
-		{
-			// **DEBUG**
-			SetStatusText("Join reply received with wrong Session ID");
-			return;						// Discard packet
-		}
 
-		// Check for JoinGranted
-		if (packet.tlMessage.tlHeader.commandType == tlcJoinGranted)
-		{
-			// Join Accepted
-			// -------------
-			// Inform Transport Layer
-			opuNetTransportLayer->OnJoinAccepted(packet);
+	opuNetTransportLayer->OnJoinAccepted(packet);
 
-			// Raise the event
-			OnJoinAccepted();
-		}
-		else
-		{
-			// Session full. Inform User
-			SetStatusText("Join Failed:  The requested game is full");
-		}
+	OnJoinAccepted();
 
-		// Reset joining game status
-		joiningGame = nullptr;
-		return;							// Packet handled
-	case tlcEchoExternalAddress:
-		// Verify packet size
-		if (packet.header.sizeOfPayload != sizeof(EchoExternalAddress)) {
-			return;						// Discard packet
-		}
-		// Verify packet came from game server **TODO**
+	joiningGame = nullptr;
+}
 
-		// Check where the information came from
-		if (packet.tlMessage.echoExternalAddress.replyPort == internalPort)
-		{
-			bReceivedInternal = true;
-		}
-		// Check if we've received a different external port than previous
-		if (ntohs(packet.tlMessage.echoExternalAddress.addr.sin_port) != externalPort)
-		{
-			if (externalPort != 0)
-			{
-				bTwoExternal = true;	// Received two distinct external ports
-			}
-		}
-		// Record external information
-		externalIp = packet.tlMessage.echoExternalAddress.addr.sin_addr;
-		if ((externalPort == 0) || (externalPort == internalPort))
-		{
-			externalPort = ntohs(packet.tlMessage.echoExternalAddress.addr.sin_port);
-		}
+void OPUNetGameSelectWnd::ReceiveJoinRefused(Packet& packet)
+{
+	if (!ReceiveJoin(packet)) {
+		return; // Discard packet
+	}
 
-		// Confine scope of `std::string text` to within a single switch case
+	SetStatusText("Join Failed:  The requested game is full");
+
+	joiningGame = nullptr;
+}
+
+bool OPUNetGameSelectWnd::ReceiveJoin(Packet& packet)
+{
+	// Verify packet size
+	if (packet.header.sizeOfPayload != sizeof(JoinReply)) {
+		return false; // Discard packet
+	}
+	// Make sure we've requested to join a game
+	if (joiningGame == nullptr)
+	{
+		SetStatusText("Unexpected Join reply received");
+		return false; // Discard packet
+	}
+	// Check the session identifier
+	if (packet.tlMessage.joinReply.sessionIdentifier != joiningGame->sessionIdentifier)
+	{
+		SetStatusText("Join reply received with wrong Session ID");

```

Comment:
I think we may also want to log a lot of the SetStatusText calls if they are updating the user on an error state.

At first I thought about adding a log call to SetStatusText, but then realized you probably do not want to log some of the status text calls, esp if they are just statusing sucessful operations.

Then I thought about adding a SetStatusTextError message that would automatically log in addition to showing the status text. But then I realized you might want to provide more details in the log message than were shown in limited space on the GUI.

So I left it alone for now for a future update.

Reply:
Hmm, that a good point, and a rather interesting idea.

The details may get a bit nuanced. Though we'll probably learn more after trying stuff out.